examples/structured_data/feature_preprocessing_layer.py

```diff
@@ -0,0 +1,224 @@
+can be mapped from a dataframe to something like tf.data.Datasets type or numpy array
+before feeding it to a Keras model.
+
+In this example we will implement feature preprocessing natively with
+Keras by implementing a custom Keras layer for all feature preprocessing steps.This layer
+utilizes Keras preprocessing layers. For constructing our model we will use the Functional API.
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+from tensorflow.keras.layers import (
+    StringLookup,
+    IntegerLookup,
+    Normalization,
+    Embedding,
+    Discretization,
+    Reshape,
+)
+import pandas as pd
+
+"""
+## Load data
+
+For this example we wil use the income dataset. This dataset contains a mixture of
+numerical and categorical features some with high cardinality.
+"""
+
+CSV_HEADER = [
+    "age",
+    "workclass",
+    "fnlwgt",
+    "education",
+    "education_num",
+    "marital_status",
+    "occupation",
+    "relationship",
+    "race",
+    "gender",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+    "native_country",
+    "income_bracket",
+]
+
+data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+
+data_frame = pd.read_csv(data_url, header=None, names=CSV_HEADER)
+labels = data_frame.pop("income_bracket")
+labels_binary = 1.0 * (labels == " >50K")
+
+print(f"Train dataset shape: {data_frame.shape}")
+
+"""
+## Split features into different feature groups
+"""
+
+NUMERIC_FEATURES = [
+    "education_num",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+    "fnlwgt",
+]
+
+CATEGORICAL_FEATURES = [
+    "workclass",
+    "marital_status",
+    "relationship",
+    "gender",
+    "education",
+    "occupation",
+]
+
+DISCRETIZE_FEATURES = ["age"]
+
+EMBEDDING_FEATURES = ["native_country"]
+
+# split data frame into different feature groups
+data_dict = {
+    "numerical_features": data_frame[NUMERIC_FEATURES].values,
+    "categorical_features": data_frame[CATEGORICAL_FEATURES].values,
+    "discretize_features": data_frame[DISCRETIZE_FEATURES].values,
+    "embedding_features": data_frame[EMBEDDING_FEATURES].values,
+}
+
+"""
+## Create Feature Preprocessing layer
+
+In this section we will implement a custom Keras layer for preprocessing. We have an adapt
+method that will update our preprocessing layer's states and apply them in the forward pass. We have
+one feature that has high cardinality and we will use embedding for this feature.
+
+"""
+
+
+class FeaturePreprocessor(tf.keras.models.Model):
+    def __init__(self, *args, **kwargs):
+        super(FeaturePreprocessor, self).__init__(*args, **kwargs)
+        # set preprocessing layers
+        self.normalizer = Normalization()
+        self.categoric_string_lookup = StringLookup()
+        self.integer_lookup = IntegerLookup(output_mode="binary")
+        self.discretizer = Discretization(num_bins=10)
+        self.embedding_string_lookup = StringLookup()
```

Comment:
You should use one layer per individual feature, don't reuse layers across multiple features